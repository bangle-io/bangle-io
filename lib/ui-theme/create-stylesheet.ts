import {
  createRawVar,
  getFromPath,
  vars,
  walkObject,
} from '@bangle.io/css-vars';
import { deepMerge, difference, intersect } from '@bangle.io/mini-js-utils';
import type { DesignTokens, RecursivePartial } from '@bangle.io/shared-types';

import {
  defaultSmallScreenOverride2,
  defaultSmallScreenOverrideDark,
  defaultSmallScreenOverrideLight,
  defaultTokensDark,
  defaultTokensLight,
} from './default-tokens';

if (typeof window !== 'undefined') {
  // This package is only meant to be used for creating stylesheets in node.js
  throw new Error('This file should not be imported in browser');
}

export const CSS_ROOT = ':root';
export const CSS_BODY = 'body';
export const CSS_SM_BODY = 'body.BU_smallscreen';
export const CSS_LIGHT_SCHEME = `.light-scheme`;
export const CSS_SM_LIGHT_SCHEME = `.light-scheme.BU_smallscreen`;
export const CSS_DARK_SCHEME = `.dark-scheme`;
export const CSS_SM_DARK_SCHEME = `.dark-scheme.BU_smallscreen`;

type CssBlocks = Record<string, string[]>;

const base = {
  [CSS_BODY]: [
    `background-color: ${vars.color.neutral.bgLayerBottom};`,
    `color: ${vars.color.neutral.text};`,
  ],
};

export type ThemeBase = RecursivePartial<
  Omit<DesignTokens, 'uid' | 'theme' | 'widescreenWidth'>
>;

function mergeWithDefaults(
  name: string,
  defaultValue: DesignTokens,
  theme: ThemeBase,
) {
  const mergedTokens = deepMerge(defaultValue, {
    theme: name,
    ...theme,
  });

  return cssVarsToStrArray(convertTokensToVars(mergedTokens));
}

export function createStyleSheetObj(
  input:
    | {
        name: string;
        type: 'single';
        theme: ThemeBase;
        smallscreenOverride?: ThemeBase;
      }
    | {
        name: string;
        type: 'dual';
        lightTheme: ThemeBase;
        darkTheme: ThemeBase;
        lightSmallscreenOverride?: ThemeBase;
        darkSmallscreenOverride?: ThemeBase;
      },
) {
  let result: CssBlocks = { ...base };

  if (input.type === 'single') {
    result = {
      ...result,
      [CSS_ROOT]: mergeWithDefaults(
        input.name,
        defaultTokensLight,
        input.theme,
      ),
    };

    let smOverride = deepMerge(
      defaultSmallScreenOverride2,
      input.smallscreenOverride || {},
    );

    result = {
      ...result,
      [CSS_SM_BODY]: createSmallScreenOverride(smOverride),
    };
  } else {
    const lightThemeVars = mergeWithDefaults(
      input.name,
      defaultTokensLight,
      input.lightTheme,
    );
    const darkThemeVars = mergeWithDefaults(
      input.name,
      defaultTokensDark,
      input.darkTheme,
    );

    const common = intersect(lightThemeVars, darkThemeVars);

    result = {
      ...result,
      [CSS_ROOT]: common,
      [CSS_LIGHT_SCHEME]: difference(lightThemeVars, common),
      [CSS_DARK_SCHEME]: difference(darkThemeVars, common),
    };

    let lightSmOverride = deepMerge(
      defaultSmallScreenOverrideLight,
      input.lightSmallscreenOverride || {},
    );

    let darkSmOverride = deepMerge(
      defaultSmallScreenOverrideDark,
      input.darkSmallscreenOverride || {},
    );

    result = {
      ...result,
      [CSS_SM_LIGHT_SCHEME]: createSmallScreenOverride(lightSmOverride),
      [CSS_SM_DARK_SCHEME]: createSmallScreenOverride(darkSmOverride),
    };
  }

  return result;
}

export function createStyleSheet(
  obj: Parameters<typeof createStyleSheetObj>[0],
): string {
  const result = createStyleSheetObj(obj);

  return Object.entries(result).reduce((prev, cur) => {
    return (
      prev +
      '\n\n' +
      wrapVarsCss(
        cur[0],
        cur[1].map((v) => '  ' + v.trim()),
      )
    );
  }, '');
}

export function createSmallScreenOverride(override: ThemeBase) {
  let result: Array<[string, string]> = [];

  walkObject(override as any, (value, path) => {
    if (typeof value === 'string') {
      result.push([createRawVar(path), value]);
    }
  });

  return cssVarsToStrArray(result).sort();
}

export function cssVarsToStrArray(vars: Array<[string, string]>) {
  return vars.map(([property, value]) => `${property}: ${value};`);
}

export function convertTokensToVars(
  _tokens: Record<string, any>,
): Array<[string, string]> {
  let tokens = _tokens;
  let result: Array<[string, string]> = [];

  walkObject(vars, (value, path) => {
    if (!value.startsWith('var(') && !value.endsWith(')')) {
      throw new Error('Invalid value ' + value);
    }

    let property = value.slice('var('.length, -1);

    let val = getFromPath(tokens, path);

    // remove uid from tokens as it does  have any use in css vars
    if (!property.endsWith('-uid') && typeof val === 'string') {
      result.push([property, val]);
    }
  });

  return result;
}

function wrapVarsCss(header: string, cssLines: string[]) {
  return `/* WARNING: This is autogenerated do not modify! */
/* stylelint-disable */
${header} {
  ${cssLines.join('\n')}
}
/* stylelint-enable */`.trim();
}
