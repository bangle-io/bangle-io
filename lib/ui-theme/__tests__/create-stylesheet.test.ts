import { createStyleSheet } from '../create-stylesheet';
import { createTokens } from '../create-tokens';

test('creates a stylesheet', () => {
  let result = createStyleSheet({
    name: 'core-theme-test',
    border: {},
    color: {},
  });
  expect(result).toMatchSnapshot();
});

test('is able to override color stylesheet', () => {
  let original = createStyleSheet({
    name: 'core-theme-test',
    border: {},
    color: {
      caution: {},
    },
  });

  expect(original).toContain('--BV-color-caution-icon: rgb(203, 93, 0);');

  let override = createStyleSheet({
    name: 'core-theme-test',
    border: {},
    color: {
      caution: {
        icon: 'red',
      },
    },
  });
  expect(override).toContain('--BV-color-caution-icon: red;');
});

describe('createTokens', () => {
  test('creates tokens correctly', () => {
    const result = createTokens({
      name: 'core-theme-test',
      border: {},
      color: {
        dark: {
          app: {
            activitybarBg: 'test-create-tokens',
          },
        },
        light: {
          caution: {},
        },
      },
    });

    expect(result).toMatchSnapshot();
  });

  test('is able to override only light', () => {
    let override = createTokens({
      name: 'core-theme-test',
      border: {},
      color: {
        dark: {
          caution: {},
        },
        light: {
          caution: {
            icon: 'light-red',
          },
        },
      },
    });

    if (!Array.isArray(override)) {
      throw new Error('Expected an array');
    }

    const lightResult = override[0];
    const darkResult = override[1];

    expect(lightResult.color.caution.icon).toContain('light-red');
    expect(darkResult.color.caution.icon).toContain('rgb(232, 116, 0)');
  });

  test('is able to app', () => {
    let override = createTokens({
      name: 'core-theme-test',
      border: {},
      color: {
        dark: {
          app: {
            activitybarBg: 'activitybar-dark-red',
          },
        },
        light: {
          caution: {},
        },
      },
    });

    if (!Array.isArray(override)) {
      throw new Error('Expected an array');
    }

    const lightResult = override[0];
    const darkResult = override[1];

    expect(darkResult.color.app.activitybarBg).toContain(
      'activitybar-dark-red',
    );
    expect(lightResult.color.app.activitybarBg).toContain('rgb(26, 32, 44)');
  });
});
