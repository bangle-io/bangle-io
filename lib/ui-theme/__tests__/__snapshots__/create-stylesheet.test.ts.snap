// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTokens creates tokens correctly 1`] = `
[
  {
    "border": {
      "radius": {
        "DEFAULT": "0.25rem",
        "lg": "0.375rem",
        "md": "0.25rem",
        "none": "0",
        "sm": "0.125rem",
        "xl": "0.5rem",
      },
      "width": {
        "DEFAULT": "1px",
        "lg": "2px",
        "md": "1px",
        "none": "0px",
      },
    },
    "color": {
      "app": {
        "activitybarBg": "rgb(26, 32, 44)",
        "activitybarText": "rgb(255, 255, 255)",
        "editorBg": "rgb(255, 255, 255)",
      },
      "caution": {
        "border": "rgb(255, 187, 99)",
        "borderStrong": "rgb(255, 160, 55)",
        "borderSubdued": "rgb(253, 210, 145)",
        "icon": "rgb(203, 93, 0)",
        "iconDisabled": "rgb(255, 187, 99)",
        "iconSubdued": "rgb(228, 111, 0)",
        "solid": "rgb(246, 133, 17)",
        "solidFaint": "rgb(253, 210, 145)",
        "solidStrong": "rgb(228, 111, 0)",
        "solidStronger": "rgb(203, 93, 0)",
        "solidSubdued": "rgb(255, 160, 55)",
        "solidText": "rgb(255, 255, 255)",
      },
      "critical": {
        "border": "rgb(255, 183, 169)",
        "borderStrong": "rgb(255, 155, 136)",
        "borderSubdued": "rgb(255, 205, 195)",
        "icon": "rgb(234, 56, 41)",
        "iconDisabled": "rgb(255, 183, 169)",
        "iconSubdued": "rgb(247, 92, 70)",
        "solid": "rgb(234, 56, 41)",
        "solidFaint": "rgb(255, 183, 169)",
        "solidStrong": "rgb(211, 21, 16)",
        "solidStronger": "rgb(180, 0, 0)",
        "solidSubdued": "rgb(247, 92, 70)",
        "solidText": "rgb(255, 255, 255)",
      },
      "neutral": {
        "bgLayerBottom": "rgb(230, 230, 230)",
        "bgLayerFloat": "rgb(253, 253, 253)",
        "bgLayerMiddle": "rgb(248, 248, 248)",
        "bgLayerTop": "rgb(255, 255, 255)",
        "border": "rgb(213, 213, 213)",
        "borderStrong": "rgb(177, 177, 177)",
        "borderSubdued": "rgb(230, 230, 230)",
        "icon": "rgb(34, 34, 34)",
        "iconDisabled": "rgb(177, 177, 177)",
        "iconSubdued": "rgb(70, 70, 70)",
        "solid": "rgb(144, 144, 144)",
        "solidFaint": "rgb(213, 213, 213)",
        "solidStrong": "rgb(109, 109, 109)",
        "solidStronger": "rgb(70, 70, 70)",
        "solidSubdued": "rgb(177, 177, 177)",
        "solidText": "rgb(255, 255, 255)",
        "text": "rgb(34, 34, 34)",
        "textDisabled": "rgb(144, 144, 144)",
        "textInverted": "rgb(255, 255, 255)",
        "textStrong": "rgb(0, 0, 0)",
        "textSubdued": "rgb(70, 70, 70)",
      },
      "positive": {
        "border": "rgb(103, 222, 168)",
        "borderStrong": "rgb(73, 204, 147)",
        "borderSubdued": "rgb(137, 236, 188)",
        "icon": "rgb(0, 143, 93)",
        "iconDisabled": "rgb(103, 222, 168)",
        "iconSubdued": "rgb(21, 164, 110)",
        "solid": "rgb(0, 143, 93)",
        "solidFaint": "rgb(103, 222, 168)",
        "solidStrong": "rgb(0, 122, 77)",
        "solidStronger": "rgb(0, 101, 62)",
        "solidSubdued": "rgb(21, 164, 110)",
        "solidText": "rgb(255, 255, 255)",
      },
      "promote": {
        "border": "rgb(150, 206, 253)",
        "borderStrong": "rgb(120, 187, 250)",
        "borderSubdued": "rgb(181, 222, 255)",
        "icon": "rgb(20, 122, 243)",
        "iconDisabled": "rgb(150, 206, 253)",
        "iconSubdued": "rgb(56, 146, 243)",
        "solid": "rgb(2, 101, 220)",
        "solidFaint": "rgb(150, 206, 253)",
        "solidStrong": "rgb(2, 101, 220)",
        "solidStronger": "rgb(0, 84, 182)",
        "solidSubdued": "rgb(20, 122, 243)",
        "solidText": "rgb(255, 255, 255)",
      },
      "secondary": {
        "border": "rgb(213, 213, 213)",
        "borderStrong": "rgb(177, 177, 177)",
        "borderSubdued": "rgb(230, 230, 230)",
        "icon": "rgb(34, 34, 34)",
        "iconDisabled": "rgb(177, 177, 177)",
        "iconSubdued": "rgb(70, 70, 70)",
        "solid": "rgb(213, 213, 213)",
        "solidFaint": "rgb(230, 230, 230)",
        "solidStrong": "rgb(230, 230, 230)",
        "solidStronger": "rgb(230, 230, 230)",
        "solidSubdued": "rgb(177, 177, 177)",
        "solidText": "rgb(34, 34, 34)",
      },
    },
    "misc": {},
    "ringWidth": {
      "DEFAULT": "1px",
      "none": "0",
    },
    "size": {
      "2xl": "42rem",
      "3xl": "48rem",
      "4xl": "56rem",
      "5xl": "64rem",
      "6xl": "72rem",
      "7xl": "80rem",
      "lg": "32rem",
      "md": "28rem",
      "prose": "65ch",
      "sm": "24rem",
      "xl": "36rem",
      "xs": "20rem",
    },
    "space": {
      "0": "0",
      "0_5": "0.125rem",
      "1": "0.25rem",
      "10": "2.5rem",
      "11": "2.75rem",
      "12": "3rem",
      "14": "3.5rem",
      "16": "4rem",
      "1_5": "0.375rem",
      "2": "0.5rem",
      "20": "5rem",
      "24": "6rem",
      "2_5": "0.625rem",
      "3": "0.75rem",
      "4": "1rem",
      "48": "12rem",
      "5": "1.25rem",
      "6": "1.5rem",
      "64": "16rem",
      "7": "1.75rem",
      "72": "18rem",
      "8": "2rem",
      "9": "2.25rem",
      "px": "1px",
    },
    "theme": "core-theme-test",
    "typography": {
      "fontFamily": {
        "mono": "ui-monospace,SFMono-Regular,Menlo,Monaco,Consolas,"Liberation Mono","Courier New",monospace",
        "sans": "ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji"",
        "serif": "ui-serif,Georgia,Cambria,"Times New Roman",Times,serif",
      },
      "text": {
        "2xl": {
          "height": "2rem",
          "size": "1.5rem",
        },
        "3xl": {
          "height": "2.25rem",
          "size": "1.875rem",
        },
        "base": {
          "height": "1.5rem",
          "size": "1rem",
        },
        "lg": {
          "height": "1.75rem",
          "size": "1.125rem",
        },
        "sm": {
          "height": "1.25rem",
          "size": "0.875rem",
        },
        "xl": {
          "height": "1.75rem",
          "size": "1.25rem",
        },
        "xs": {
          "height": "1rem",
          "size": "0.75rem",
        },
      },
    },
    "uid": "design-token::core-theme-test-light",
    "widescreenWidth": "759px",
  },
  {
    "border": {
      "radius": {
        "DEFAULT": "0.25rem",
        "lg": "0.375rem",
        "md": "0.25rem",
        "none": "0",
        "sm": "0.125rem",
        "xl": "0.5rem",
      },
      "width": {
        "DEFAULT": "1px",
        "lg": "2px",
        "md": "1px",
        "none": "0px",
      },
    },
    "color": {
      "app": {
        "activitybarBg": "test-create-tokens",
        "activitybarText": "rgb(235, 235, 235)",
        "editorBg": "rgb(0, 0, 0)",
      },
      "caution": {
        "border": "rgb(210, 98, 0)",
        "borderStrong": "rgb(232, 116, 0)",
        "borderSubdued": "rgb(186, 82, 0)",
        "icon": "rgb(232, 116, 0)",
        "iconDisabled": "rgb(138, 55, 0)",
        "iconSubdued": "rgb(210, 98, 0)",
        "solid": "rgb(249, 137, 23)",
        "solidFaint": "rgb(115, 43, 0)",
        "solidStrong": "rgb(255, 162, 59)",
        "solidStronger": "rgb(255, 188, 102)",
        "solidSubdued": "rgb(235, 235, 235)",
        "solidText": "rgb(235, 235, 235)",
      },
      "critical": {
        "border": "rgb(238, 67, 49)",
        "borderStrong": "rgb(249, 99, 76)",
        "borderSubdued": "rgb(221, 33, 24)",
        "icon": "rgb(249, 99, 76)",
        "iconDisabled": "rgb(167, 0, 0)",
        "iconSubdued": "rgb(238, 67, 49)",
        "solid": "rgb(249, 99, 76)",
        "solidFaint": "rgb(110, 0, 0)",
        "solidStrong": "rgb(255, 129, 107)",
        "solidStronger": "rgb(255, 158, 140)",
        "solidSubdued": "rgb(238, 67, 49)",
        "solidText": "rgb(235, 235, 235)",
      },
      "neutral": {
        "bgLayerBottom": "rgb(0, 0, 0)",
        "bgLayerFloat": "rgb(29, 29, 29)",
        "bgLayerMiddle": "rgb(14, 14, 14)",
        "bgLayerTop": "rgb(29, 29, 29)",
        "border": "rgb(75, 75, 75)",
        "borderStrong": "rgb(106, 106, 106)",
        "borderSubdued": "rgb(48, 48, 48)",
        "icon": "rgb(235, 235, 235)",
        "iconDisabled": "rgb(106, 106, 106)",
        "iconSubdued": "rgb(208, 208, 208)",
        "solid": "rgb(106, 106, 106)",
        "solidFaint": "rgb(48, 48, 48)",
        "solidStrong": "rgb(75, 75, 75)",
        "solidStronger": "rgb(48, 48, 48)",
        "solidSubdued": "rgb(141, 141, 141)",
        "solidText": "rgb(235, 235, 235)",
        "text": "rgb(235, 235, 235)",
        "textDisabled": "rgb(141, 141, 141)",
        "textInverted": "rgb(0, 0, 0)",
        "textStrong": "rgb(255, 255, 255)",
        "textSubdued": "rgb(208, 208, 208)",
      },
      "positive": {
        "border": "rgb(0, 149, 98)",
        "borderStrong": "rgb(28, 168, 114)",
        "borderSubdued": "rgb(0, 130, 82)",
        "icon": "rgb(28, 168, 114)",
        "iconDisabled": "rgb(0, 93, 57)",
        "iconSubdued": "rgb(0, 149, 98)",
        "solid": "rgb(52, 187, 132)",
        "solidFaint": "rgb(0, 93, 57)",
        "solidStrong": "rgb(75, 205, 149)",
        "solidStronger": "rgb(103, 222, 168)",
        "solidSubdued": "rgb(28, 168, 114)",
        "solidText": "rgb(235, 235, 235)",
      },
      "promote": {
        "border": "rgb(227, 69, 137)",
        "borderStrong": "rgb(241, 97, 156)",
        "borderSubdued": "rgb(209, 43, 114)",
        "icon": "rgb(241, 97, 156)",
        "iconDisabled": "rgb(161, 0, 78)",
        "iconSubdued": "rgb(227, 69, 137)",
        "solid": "rgb(241, 97, 156)",
        "solidFaint": "rgb(133, 0, 65)",
        "solidStrong": "rgb(252, 124, 173)",
        "solidStronger": "rgb(255, 152, 191)",
        "solidSubdued": "rgb(227, 69, 137)",
        "solidText": "rgb(235, 235, 235)",
      },
      "secondary": {
        "border": "rgb(75, 75, 75)",
        "borderStrong": "rgb(106, 106, 106)",
        "borderSubdued": "rgb(48, 48, 48)",
        "icon": "rgb(235, 235, 235)",
        "iconDisabled": "rgb(106, 106, 106)",
        "iconSubdued": "rgb(208, 208, 208)",
        "solid": "rgb(75, 75, 75)",
        "solidFaint": "rgb(48, 48, 48)",
        "solidStrong": "rgb(48, 48, 48)",
        "solidStronger": "rgb(48, 48, 48)",
        "solidSubdued": "rgb(106, 106, 106)",
        "solidText": "rgb(208, 208, 208)",
      },
    },
    "misc": {},
    "ringWidth": {
      "DEFAULT": "1px",
      "none": "0",
    },
    "size": {
      "2xl": "42rem",
      "3xl": "48rem",
      "4xl": "56rem",
      "5xl": "64rem",
      "6xl": "72rem",
      "7xl": "80rem",
      "lg": "32rem",
      "md": "28rem",
      "prose": "65ch",
      "sm": "24rem",
      "xl": "36rem",
      "xs": "20rem",
    },
    "space": {
      "0": "0",
      "0_5": "0.125rem",
      "1": "0.25rem",
      "10": "2.5rem",
      "11": "2.75rem",
      "12": "3rem",
      "14": "3.5rem",
      "16": "4rem",
      "1_5": "0.375rem",
      "2": "0.5rem",
      "20": "5rem",
      "24": "6rem",
      "2_5": "0.625rem",
      "3": "0.75rem",
      "4": "1rem",
      "48": "12rem",
      "5": "1.25rem",
      "6": "1.5rem",
      "64": "16rem",
      "7": "1.75rem",
      "72": "18rem",
      "8": "2rem",
      "9": "2.25rem",
      "px": "1px",
    },
    "theme": "core-theme-test",
    "typography": {
      "fontFamily": {
        "mono": "ui-monospace,SFMono-Regular,Menlo,Monaco,Consolas,"Liberation Mono","Courier New",monospace",
        "sans": "ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji"",
        "serif": "ui-serif,Georgia,Cambria,"Times New Roman",Times,serif",
      },
      "text": {
        "2xl": {
          "height": "2rem",
          "size": "1.5rem",
        },
        "3xl": {
          "height": "2.25rem",
          "size": "1.875rem",
        },
        "base": {
          "height": "1.5rem",
          "size": "1rem",
        },
        "lg": {
          "height": "1.75rem",
          "size": "1.125rem",
        },
        "sm": {
          "height": "1.25rem",
          "size": "0.875rem",
        },
        "xl": {
          "height": "1.75rem",
          "size": "1.25rem",
        },
        "xs": {
          "height": "1rem",
          "size": "0.75rem",
        },
      },
    },
    "uid": "design-token::core-theme-test-dark",
    "widescreenWidth": "759px",
  },
]
`;

exports[`creates a stylesheet 1`] = `
"/* WARNING: This is autogenerated do not modify! */
/* stylelint-disable */
body {
    background-color: var(--BV-color-neutral-bgLayerBottom);
  color: var(--BV-color-neutral-text);
}
/* stylelint-enable */

/* WARNING: This is autogenerated do not modify! */
/* stylelint-disable */
:root {
    --BV-theme: core-theme-test;
  --BV-border-radius-DEFAULT: 0.25rem;
  --BV-border-radius-none: 0;
  --BV-border-radius-sm: 0.125rem;
  --BV-border-radius-md: 0.25rem;
  --BV-border-radius-lg: 0.375rem;
  --BV-border-radius-xl: 0.5rem;
  --BV-border-width-DEFAULT: 1px;
  --BV-border-width-none: 0px;
  --BV-border-width-lg: 2px;
  --BV-border-width-md: 1px;
  --BV-widescreenWidth: 759px;
  --BV-color-app-activitybarBg: rgb(26, 32, 44);
  --BV-color-app-activitybarText: rgb(255, 255, 255);
  --BV-color-app-editorBg: rgb(255, 255, 255);
  --BV-color-neutral-bgLayerTop: rgb(255, 255, 255);
  --BV-color-neutral-bgLayerMiddle: rgb(248, 248, 248);
  --BV-color-neutral-bgLayerBottom: rgb(230, 230, 230);
  --BV-color-neutral-bgLayerFloat: rgb(253, 253, 253);
  --BV-color-neutral-textDisabled: rgb(144, 144, 144);
  --BV-color-neutral-textSubdued: rgb(70, 70, 70);
  --BV-color-neutral-text: rgb(34, 34, 34);
  --BV-color-neutral-textStrong: rgb(0, 0, 0);
  --BV-color-neutral-textInverted: rgb(255, 255, 255);
  --BV-color-neutral-iconDisabled: rgb(177, 177, 177);
  --BV-color-neutral-iconSubdued: rgb(70, 70, 70);
  --BV-color-neutral-icon: rgb(34, 34, 34);
  --BV-color-neutral-solid: rgb(144, 144, 144);
  --BV-color-neutral-solidText: rgb(255, 255, 255);
  --BV-color-neutral-solidStrong: rgb(109, 109, 109);
  --BV-color-neutral-solidStronger: rgb(70, 70, 70);
  --BV-color-neutral-solidSubdued: rgb(177, 177, 177);
  --BV-color-neutral-solidFaint: rgb(213, 213, 213);
  --BV-color-neutral-borderSubdued: rgb(230, 230, 230);
  --BV-color-neutral-border: rgb(213, 213, 213);
  --BV-color-neutral-borderStrong: rgb(177, 177, 177);
  --BV-color-secondary-iconDisabled: rgb(177, 177, 177);
  --BV-color-secondary-iconSubdued: rgb(70, 70, 70);
  --BV-color-secondary-icon: rgb(34, 34, 34);
  --BV-color-secondary-solid: rgb(213, 213, 213);
  --BV-color-secondary-solidText: rgb(34, 34, 34);
  --BV-color-secondary-solidStrong: rgb(230, 230, 230);
  --BV-color-secondary-solidStronger: rgb(230, 230, 230);
  --BV-color-secondary-solidSubdued: rgb(177, 177, 177);
  --BV-color-secondary-solidFaint: rgb(230, 230, 230);
  --BV-color-secondary-borderSubdued: rgb(230, 230, 230);
  --BV-color-secondary-border: rgb(213, 213, 213);
  --BV-color-secondary-borderStrong: rgb(177, 177, 177);
  --BV-color-promote-iconDisabled: rgb(150, 206, 253);
  --BV-color-promote-iconSubdued: rgb(56, 146, 243);
  --BV-color-promote-icon: rgb(20, 122, 243);
  --BV-color-promote-solid: rgb(2, 101, 220);
  --BV-color-promote-solidText: rgb(255, 255, 255);
  --BV-color-promote-solidStrong: rgb(2, 101, 220);
  --BV-color-promote-solidStronger: rgb(0, 84, 182);
  --BV-color-promote-solidSubdued: rgb(20, 122, 243);
  --BV-color-promote-solidFaint: rgb(150, 206, 253);
  --BV-color-promote-borderSubdued: rgb(181, 222, 255);
  --BV-color-promote-border: rgb(150, 206, 253);
  --BV-color-promote-borderStrong: rgb(120, 187, 250);
  --BV-color-caution-iconDisabled: rgb(255, 187, 99);
  --BV-color-caution-iconSubdued: rgb(228, 111, 0);
  --BV-color-caution-icon: rgb(203, 93, 0);
  --BV-color-caution-solid: rgb(246, 133, 17);
  --BV-color-caution-solidText: rgb(255, 255, 255);
  --BV-color-caution-solidStrong: rgb(228, 111, 0);
  --BV-color-caution-solidStronger: rgb(203, 93, 0);
  --BV-color-caution-solidSubdued: rgb(255, 160, 55);
  --BV-color-caution-solidFaint: rgb(253, 210, 145);
  --BV-color-caution-borderSubdued: rgb(253, 210, 145);
  --BV-color-caution-border: rgb(255, 187, 99);
  --BV-color-caution-borderStrong: rgb(255, 160, 55);
  --BV-color-critical-iconDisabled: rgb(255, 183, 169);
  --BV-color-critical-iconSubdued: rgb(247, 92, 70);
  --BV-color-critical-icon: rgb(234, 56, 41);
  --BV-color-critical-solid: rgb(234, 56, 41);
  --BV-color-critical-solidText: rgb(255, 255, 255);
  --BV-color-critical-solidStrong: rgb(211, 21, 16);
  --BV-color-critical-solidStronger: rgb(180, 0, 0);
  --BV-color-critical-solidSubdued: rgb(247, 92, 70);
  --BV-color-critical-solidFaint: rgb(255, 183, 169);
  --BV-color-critical-borderSubdued: rgb(255, 205, 195);
  --BV-color-critical-border: rgb(255, 183, 169);
  --BV-color-critical-borderStrong: rgb(255, 155, 136);
  --BV-color-positive-iconDisabled: rgb(103, 222, 168);
  --BV-color-positive-iconSubdued: rgb(21, 164, 110);
  --BV-color-positive-icon: rgb(0, 143, 93);
  --BV-color-positive-solid: rgb(0, 143, 93);
  --BV-color-positive-solidText: rgb(255, 255, 255);
  --BV-color-positive-solidStrong: rgb(0, 122, 77);
  --BV-color-positive-solidStronger: rgb(0, 101, 62);
  --BV-color-positive-solidSubdued: rgb(21, 164, 110);
  --BV-color-positive-solidFaint: rgb(103, 222, 168);
  --BV-color-positive-borderSubdued: rgb(137, 236, 188);
  --BV-color-positive-border: rgb(103, 222, 168);
  --BV-color-positive-borderStrong: rgb(73, 204, 147);
  --BV-ringWidth-DEFAULT: 1px;
  --BV-ringWidth-none: 0;
  --BV-size-xs: 20rem;
  --BV-size-sm: 24rem;
  --BV-size-md: 28rem;
  --BV-size-lg: 32rem;
  --BV-size-xl: 36rem;
  --BV-size-2xl: 42rem;
  --BV-size-3xl: 48rem;
  --BV-size-4xl: 56rem;
  --BV-size-5xl: 64rem;
  --BV-size-6xl: 72rem;
  --BV-size-7xl: 80rem;
  --BV-size-prose: 65ch;
  --BV-space-0: 0;
  --BV-space-1: 0.25rem;
  --BV-space-2: 0.5rem;
  --BV-space-3: 0.75rem;
  --BV-space-4: 1rem;
  --BV-space-5: 1.25rem;
  --BV-space-6: 1.5rem;
  --BV-space-7: 1.75rem;
  --BV-space-8: 2rem;
  --BV-space-9: 2.25rem;
  --BV-space-10: 2.5rem;
  --BV-space-11: 2.75rem;
  --BV-space-12: 3rem;
  --BV-space-14: 3.5rem;
  --BV-space-16: 4rem;
  --BV-space-20: 5rem;
  --BV-space-24: 6rem;
  --BV-space-48: 12rem;
  --BV-space-64: 16rem;
  --BV-space-72: 18rem;
  --BV-space-0_5: 0.125rem;
  --BV-space-1_5: 0.375rem;
  --BV-space-2_5: 0.625rem;
  --BV-space-px: 1px;
  --BV-typography-fontFamily-sans: ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";
  --BV-typography-fontFamily-serif: ui-serif,Georgia,Cambria,"Times New Roman",Times,serif;
  --BV-typography-fontFamily-mono: ui-monospace,SFMono-Regular,Menlo,Monaco,Consolas,"Liberation Mono","Courier New",monospace;
  --BV-typography-text-xl-height: 1.75rem;
  --BV-typography-text-xl-size: 1.25rem;
  --BV-typography-text-lg-height: 1.75rem;
  --BV-typography-text-lg-size: 1.125rem;
  --BV-typography-text-base-height: 1.5rem;
  --BV-typography-text-base-size: 1rem;
  --BV-typography-text-sm-height: 1.25rem;
  --BV-typography-text-sm-size: 0.875rem;
  --BV-typography-text-xs-height: 1rem;
  --BV-typography-text-xs-size: 0.75rem;
  --BV-typography-text-2xl-height: 2rem;
  --BV-typography-text-2xl-size: 1.5rem;
  --BV-typography-text-3xl-height: 2.25rem;
  --BV-typography-text-3xl-size: 1.875rem;
}
/* stylelint-enable */"
`;
